src/ui/annotation_schema_tab.ts

```diff
@@ -526,35 +528,82 @@ class AnnotationUIProperty extends RefCounted {
     this.defaultValueCell = cell;
     return cell;
   }
+
   private createInputElement(
     config: InputConfig,
     numberConfig?: NumberConfig,
-  ): HTMLInputElement {
+  ): HTMLInputElement | HTMLTextAreaElement {
     const readonly = this.readonly;
-    const input =
-      config.type === "number"
-        ? createBoundedNumberInputElement(
-            {
-              inputValue: config.inputValue as number,
-              className: config.className,
-              numDecimals: config.decimals,
-              readonly,
-            },
-            numberConfig,
-          )
-        : document.createElement("input");
+
+    if (config.type === "number") {
+      const input = createBoundedNumberInputElement(
+        {
+          inputValue: config.inputValue as number,
+          className: config.className,
+          numDecimals: config.decimals,
+          readonly,
+        },
+        numberConfig,
+      );
+      input.dataset.readonly = String(readonly);
+      input.disabled = readonly;
+      return input;
+    }
+
+    if (config.type === "text") {
+      if (config.useTextarea) {
+        const textarea = document.createElement("textarea");
+        textarea.dataset.readonly = String(readonly);
+        textarea.disabled = readonly;
+
+        if (config.className) textarea.classList.add(config.className);
+        textarea.value = String(config.inputValue || "");
+        textarea.autocomplete = "off";
+        textarea.spellcheck = false;
+
+        const resizeTextarea = () => {
+          textarea.style.height = 'auto';
+          textarea.style.height = `${textarea.scrollHeight}px`;
+        };
+
+        resizeTextarea();
+
+        textarea.addEventListener('input', resizeTextarea);
+
+        if (typeof ResizeObserver !== 'undefined') {
+          const observer = new ResizeObserver(resizeTextarea);
+          observer.observe(textarea);
+        }
+        
+        return textarea;
+      } else {
+        const input = document.createElement("input");
+        input.type = "text";
+        input.dataset.readonly = String(readonly);
```

Comment:
I think it might be nice if instead of repeating the setup for both non-text-area and text area inputs, you do something more like (rough names here):

```
// The logic dependent on the type happens here
const input = useTextArea ? createTextAreaElement() : createTextInputElement()
//and then do the common things for both
if (config.className) input.classList.add(config.className);
        input.value = String(config.inputValue || "");
        input.autocomplete = "off";
        input.spellcheck = false;
        input.disabled = readonly;
        input.dataset.readonly = String(readonly);
        //etc

```
that way it's clear what is logic specific to each branch of the if, and which is general logic